pypdf/_page.py

```diff
@@ -2149,6 +2151,48 @@ def annotations(self, value: Optional[ArrayObject]) -> None:
         else:
             self[NameObject("/Annots")] = value
 
+    def add_action(
+            self,
+            event: Literal["O", "C"] = "O",
+            action_type: Literal["JavaScript"] = "JavaScript",
+            action: str = ""
+        ) -> None:
+        r"""
+        Add action which will launch on the open or close event of this
+        page.
+
+        Args:
+            event: "/O" or "/C", for open or close action respectively.
+            action_type: "JavaScript" is currently the only available action type.
+            action: Your JavaScript.
+
+        >>> output.add_action("/O", "JavaScript", 'app.alert("This is page " + this.pageNum);')
+        # Example: This will display the page number when the page is opened.
+        >>> output.add_action("/C", "JavaScript", 'app.alert("This is page " + this.pageNum);')
+        # Example: This will display the page number when the page is closed.
+
+        Note that this will replace any existing open or close event on this page.
```

Comment:
These limitations still sound like we should further refine the public API. `add_action` should add the new action in a defined manner, not overwrite existing ones - especially when the specification supports having both events defined on the page, but our API design only allows one.

Reply:
You can of course open a discussion, but with JavaScript being ignored by many PDF readers and (at least from my experience) limited adoption, I guess there will not be much feedback.

This is one of the reasons why we usually ask to open issues first to discuss new APIs, although we might have missed this there as well and additional feedback would most likely have been sparse as well.